src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SubSourceLogic.cs

```diff
@@ -450,191 +565,228 @@ public SubSourcePromiseControl(SourceShape<(
-                                  IMessageBuilder<K, V, TMessage> messageBuilder,
-                                  Decider decider,
-                                  int actorNumber)
-            {
-                _topicPartition = topicPartition;
-                _consumerActor = consumerActor;
-                _subSourceStartedCallback = subSourceStartedCallback;
-                _subSourceCancelledCallback = subSourceCancelledCallback;
-                _messageBuilder = messageBuilder;
-                _decider = decider;
-                _actorNumber = actorNumber;
+        public SubSourceStage(TopicPartition topicPartition, IActorRef consumerActor,
+            Action<SubSourceStageLogicControl> subSourceStartedCallback,
+            Action<(TopicPartition, ISubSourceCancellationStrategy)> subSourceCancelledCallback,
+            Decider decider,
+            int actorNumber, ISubSourceStageLogicFactory<K, V, TMessage> subSourceStageLogicFactory)
+        {
+            _topicPartition = topicPartition;
+            _consumerActor = consumerActor;
+            _subSourceStartedCallback = subSourceStartedCallback;
+            _subSourceCancelledCallback = subSourceCancelledCallback;
+            _decider = decider;
+            _actorNumber = actorNumber;
+            _subSourceStageLogicFactory = subSourceStageLogicFactory;
+
+            Out = new Outlet<TMessage>("out");
+            Shape = new SourceShape<TMessage>(Out);
+        }
 
-                Out = new Outlet<TMessage>("out");
-                Shape = new SourceShape<TMessage>(Out);
-            }
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes)
+        {
+            return _subSourceStageLogicFactory.Create(Shape, _topicPartition, _consumerActor, 
+                _subSourceStartedCallback, _subSourceCancelledCallback, _actorNumber);
+        }
+    }
 
-            protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes)
+    /// <summary>
+    /// INTERNAL API
+    ///
+    /// A <see cref="SubSourceStageLogic{K,V,TMessage}"/> is the <see cref="GraphStageLogic"/>
+    /// of a <see cref="SubSourceStage{K,V,TMessage}"/>.
+    ///
+    /// This emits the actual Kafka messages downstream.
+    /// </summary>
+    internal abstract class SubSourceStageLogic<K, V, TMessage> : GraphStageLogic
+    {
+        private readonly SourceShape<TMessage> _shape;
+        private readonly TopicPartition _topicPartition;
+        private readonly IActorRef _consumerActor;
+        private readonly int _actorNumber;
+        private readonly IMessageBuilder<K, V, TMessage> _messageBuilder;
+        private readonly Action<SubSourceStageLogicControl> _subSourceStartedCallback;
+        private readonly KafkaConsumerActorMetadata.Internal.RequestMessages _requestMessages;
+        private bool _requested = false;
+        private StageActor _subSourceActor = null!;
+        private Queue<ConsumeResult<K, V>> _buffer = new();
```

Comment:
Just like with the non-partitioned stages: we never needed a `ConcurrentQueue<T>` - a regular old `Queue<T>` will work just fine.